frontend/catalyst/pennylane_extensions.py

```diff
@@ -522,6 +523,27 @@ def _check_grad_params(
     return GradParams(method, scalar_out, h, argnum)
 
 
+def _unflatten_derivatives(results, out_tree, argnum, num_results):
+    """Unflatten the flat list of derivatives results given the out tree."""
+    num_trainable_params = len(argnum) if isinstance(argnum, list) else 1
+    results = tuple(
+        tuple(results[i * num_trainable_params : i * num_trainable_params + num_trainable_params])
+        for i in range(0, num_results)
+    )
+    # In jax,  argnums=[int] wraps single derivatives in a tuple
+    if (
+        (isinstance(argnum, int) or argnum is None)
+        and num_trainable_params == 1
+        and num_results != 1
+    ):
+        results = tuple(r[0] if len(r) == 1 else r for r in results)
```

Comment:
I find these lines very difficult to follow, maybe there is a way to refactor them or explain what they are doing?

Reply:
This is doing:

```
flat list of derivatives -> structure of the jacobian (tuple of tuple if multiple return and multiple parameters)
```